packages/web/components/pool-detail/concentrated.tsx

```diff
@@ -86,19 +86,81 @@ export const ConcentratedLiquidityPool: FunctionComponent<{ poolId: string }> =
       ? pool.concentratedLiquidityPoolInfo.currentPrice
       : undefined;
 
-    const userPositions = osmosisQueries.queryAccountsPositions.get(
-      account?.address ?? ""
-    ).positions;
+    const userPositionsInPool = osmosisQueries.queryAccountsPositions
+      .get(account?.address ?? "")
+      .positionsInPool(poolId);
 
-    const userPositionsInPool = userPositions.filter(
-      (position) => position.poolId === poolId
-    );
     const userHasPositionInPool = userPositionsInPool.length > 0;
 
-    const rewardedPositions = userPositions.filter(
+    const rewardedPositions = userPositionsInPool.filter(
       (position) => position.hasClaimableRewards
     );
 
+    const onClickCollectAllRewards = () => {
+      const fiat = priceStore.getFiatCurrency(priceStore.defaultVsCurrency);
+
+      if (!account) throw new Error("No account");
+
+      const rewardAmountUSD = rewardedPositions.reduce(
```

Comment:
Two quick thoughts on this algorithm, 

1. We have nested `reduce` functions which causes a quadratic time complexity. Is it possible to restructure `rewardedPositions` so that we can accomplish this with a linear time complexity? I'm not sure what the structure of `rewardedPositions` is, otherwise I would suggest something. Let me know if you want to pair in case I'm not being clear
2. If we can't avoid the nested loop, maybe we should consider pulling out the inner `reduce` to be it's own named function just so this is easier to read. Something like this:
```
const calculateTotalValue = (rewards, fiat) => {
    if (!Array.isArray(rewards) || !rewards.length || !fiat) return 0;

    return rewards.reduce((sum, asset) => {
        const price = priceStore.calculatePrice(asset);
        return sum + (price ? Number(price.toDec().toString()) : 0);
    }, 0);
};

const rewardAmountUSD = rewardedPositions.reduce((acc, { totalClaimableRewards }) => {
    return acc + calculateTotalValue(totalClaimableRewards, fiat);
}, 0);
```

Reply:
I recognize that this perf issue already existed before, but I'm just noticing it now I guess 🙃

Reply:
In response:

`[positions: [coins]]`

1. I think we have a 2d array, or tree with height of 2, that needs to be traversed here (see above): an array of positions > then for each position an array of coin rewards. 90%+ of positions will have only 2 coins as rewards; the coins that the pool is comprised of (as users swap those coins, tiny amounts are left in positions that are used for the swap as the reward). Then, if the pool's being externally incentivized, that will perhaps add one additional coin in the remaining 10% of cases if that coin is different from the coins in the pool.  With that knowledge, I wasn't really thinking about performance as most users will have a few positions; my guess is it'd only get noticeable if you had 100s of positions (we'd have performance concerns elsewhere in that case 😅)
2. I agree, we can make this code easier to scan with some extraction 👍